firmware-management/mbl-watchdog-init/mbl-watchdog-init.c

```diff
@@ -0,0 +1,210 @@
+// Error handling and logging convenience functions
+int is_err(const int ret)
+{
+    return ret == -1;
+}
+
+// Uses errno to determine the last error code.
+// Must be called immediately after the call which could throw an error
+// so errno doesn't change.
+void log_error(const char * const message)
+{
+    static const char * const prefix = "WATCHDOG-ERROR:";
+    const int error_num = errno;
+    const char* const err_msg = strerror(error_num);
+
+    fprintf(stderr, "%s %s %s\n", prefix, message, err_msg);
+}
+
+
+void log_warning(const char * const message)
+{
+    static const char * const prefix = "WATCHDOG-WARNING:";
+    fprintf(stderr, "%s %s\n", prefix, message);
+}
+
+
+void log_info(const char * const message)
+{
+    static const char * const prefix = "WATCHDOG-INFO:";
+    fprintf(stderr, "%s %s\n", prefix, message);
+}
+
+//============================================================================
+// Watchdog functions
+int set_watchdog_timeout(const int watchdog_fd, int * const timeout)
+{
+    return ioctl(watchdog_fd, WDIOC_SETTIMEOUT, timeout);
+}
+
+
+void print_last_boot_reason(const int boot_status)
+{
+    switch (boot_status)
+    {
+        case WDIOF_OVERHEAT:
+            log_info("The last reboot was caused by the CPU overheating.");
+        case WDIOF_CARDRESET:
+            log_info("The last reboot was caused by a watchdog reset.");
+        case WDIOF_FANFAULT:
+            log_info("The last reboot was because a system fan monitored by the watchdog card failed.");
+        case WDIOF_EXTERN1:
+            log_info("The last reboot was because external monitoring relay/source 1 was triggered.");
+        case WDIOF_EXTERN2:
+            log_info("The last reboot was because external monitoring relay/source 2 was triggered.");
+        case WDIOF_POWERUNDER:
+            log_info("The last reboot was due to the machine showing an undervoltage status.");
+        case WDIOF_POWEROVER:
+            log_info("The last reboot was due to the machine showing an overvoltage status.");
+    }
+}
+
+
+int get_last_boot_status(const int watchdog_fd)
+{
+    int flags;
+
+    const int gs_ret = ioctl(watchdog_fd, WDIOC_GETBOOTSTATUS, &flags);
+    if (is_err(gs_ret))
+    {
+        return gs_ret;
+    }
+
+    print_last_boot_reason(flags);
+
+    return 0;
+}
+
+
+//============================================================================
+// Convert a string containing only digits to an integer. Exits on errors.
+int numeric_string_to_int(const char* const str)
+{
+    errno = 0;
+    char *endptr;
+    const long int res = strtol(str, &endptr, 0);
+
+    if ((errno == ERANGE && (res == LONG_MAX || res == LONG_MIN))
+           || (errno != 0 && res == 0))
+    {
+        log_error("strtol failed");
+        exit(EXIT_FAILURE);
+    }
+
+    if (res < 0 || res > INT_MAX)
+    {
+        log_error("Integer not in valid range. Expecting a positive integer with maximum size MAX_INT");
```

Comment:
`MAX_INT` -> `INT_MAX`. 
You haven't checked that `res` is positive, just that it's non-negative. Is zero a valid value? If so:
```C
    if (res < 0 || res > INT_MAX)
    {
        log_error("Integer not in valid range. Expecting a non-negative integer with maximum size INT_MAX");
```
Otherwise:
```C
    if (res <= 0 || res > INT_MAX)
    {
        log_error("Integer not in valid range. Expecting a positive integer with maximum size INT_MAX");
```

I suggest you change `int` to `nonnegative_int` or `positive_int` in the function name.

Reply:
Alternatively, you could ignore whether `res` is negative or zero and just check against `INT_MIN`. The caller would have to do extra checks if they need something more constrained than a valid `int` value, but the only caller so far already does that.

Reply:
8dd89775258c82300ec5d4eea7836f47c64a1584 - checks `res` is positive and isn't greater than `INT_MAX`